src/drivers/intel/dmic.c

```diff
@@ -579,9 +578,19 @@ static int select_mode(struct dai *dai,
 	else
 		mfir = modes->mfir_b;
 
-	mmin = find_min_int16(mfir, modes->num_of_modes);
-	count = find_equal_int16(idx, mfir, mmin, modes->num_of_modes, 0);
-	n = idx[count - 1];
+	/* Search fir_list[] decimation factors from start towards end. The found
+	 * last configuration entry with searched decimation factor will be used.
+	 */
+	count = 0;
+	for (n = 0; fir_list[n]; n++) {
+		count = find_equal_int16(idx, mfir, fir_list[n]->decim_factor,
+					 modes->num_of_modes, 0);
+		if (count > 0)
+			break;
+	}
+
+	assert(count); /* Should not happen. */
```

Comment:
Can we have an trace error here instead and return an error to make this survivavble.

Reply:
Sure, I changed this just to error trace and -EINVAL return. Only a developer doing changes to to decimation ratios selection should be able to get this error. I tried to force this to happen, seems like the DMA trace from topology load phase does not show this. But there's clean error in dmesg at least.